components/eam/src/chemistry/modal_aero/modal_aero_convproc.F90

```diff
@@ -1786,12 +1468,556 @@ subroutine ma_convproc_tend(                                           &
+   integer,  intent(in) :: iconvtype            ! 1=deep, 2=uw shallow
+   integer,  intent(in) :: kk                   ! vertical level index
+   real(r8), intent(in) :: mu_i(pverp)          ! mu at current i (note pverp dimension) [mb/s]
+   real(r8), intent(in) :: cldfrac_i(pver)      ! cldfrac at current icol (with adjustments) [fraction]
+   real(r8), intent(in) :: rhoair_i(pver)       ! air density at current i [kg/m3]
+   real(r8), intent(in) :: zmagl(pver)          ! height above surface [m]
+   real(r8), intent(inout)      :: wup(pver)    ! mean updraft vertical velocity at current level updraft [m/s]
+
+   integer      :: kp1                  ! kk + 1
+   real(r8)     :: zkm                  ! height above surface [km]
+   real(r8), parameter  :: w_peak = 4.0_r8      ! pre-defined peak updraft [m/s]
+
+! shallow - wup = (mup in kg/m2/s) / [rhoair * (updraft area)]
+    if (iconvtype /= 1) then
+       wup(kk) = (mu_i(kp1) + mu_i(kk))*0.5_r8*hund_ovr_g &
+                      / (rhoair_i(kk) * (cldfrac_i(kk)*0.5_r8))
+       wup(kk) = max( 0.1_r8, wup(kk) )
+
+! deep - the above method overestimates updraft area and underestimate wup
+!    the following is based lemone and zipser (j atmos sci, 1980, p. 2455)
+!    peak updraft (= 4 m/s) is sort of a "grand median" from their GATE data
+!       and Thunderstorm Project data which they also show
+!    the vertical profile shape is a crude fit to their median updraft profile
+   else
+       zkm = zmagl(kk)*1.0e-3
+       if (zkm .ge. 1.0) then
+           wup(kk) = w_peak*((zkm/w_peak)**0.21_r8)
+       else
+           wup(kk) = 2.9897_r8*(zkm**0.5_r8)
+       endif
+       wup(kk) = min_max_bound(0.1_r8, w_peak, wup(kk))
+   endif
+
+   end subroutine compute_wup
+
+!====================================================================================
+   subroutine compute_activation_tend(                          &
+                        icol,   kk,     clw_cut,        f_ent,  & ! in
+                        cldfrac_i,      rhoair_i,               & ! in
+                        dt_u,   wup,    icwmr,          t,      & ! in
+                        kactcnt,        kactfirst,              & ! inout
+                        conu,           dconudt_activa          ) ! inout
+!-----------------------------------------------------------------------
+! aerosol activation - method 2
+!    when kactcnt=1 (first/lowest layer with cloud water)
+!       apply "primary" activatation to the entire updraft
+!    when kactcnt>1
+!       apply secondary activatation to the entire updraft
+!       do this for all levels above cloud base (even if completely glaciated)
+!          (this is something for sensitivity testing)
+!-----------------------------------------------------------------------
+   use ppgrid, only: pver, pverp, pcols
+   use constituents, only: pcnst
+
+   ! cloudborne aerosol, so the arrays are dimensioned with pcnst_extd = pcnst*2
+   integer,  parameter  :: pcnst_extd = pcnst*2
```

Comment:
We can leave it as is now but `pcnst_extd` logic can be hard to follow. If it is easy, I would like to see if we can explicitly use `pcnst` and treat both interstitial and cloud borne separately.